### PR TITLE
Fixing metric names, tags, config and tests

### PR DESCRIPTION
**Pull request by ravirajj**
- The metrics are identified/key'ed by their name and tags combination.
- Removing redundant `env` and `service` from metric name, since every metric has the `env` and `service` tags.
- Removing redundant `per-workers` and `all-workers` from metric name, and as per M3 best practices, creating separate all-host and per-host metrics to reduce metric cardinality when querying metrics for all hosts.
- Removing redundant `metrics.m3.service` and `metrics.m3.flushInterval` config.
- Setting defaults in config/production.json instead of hardcoding it in the implementation.
- Using combined name and tag keys to check if metrics are emitted correctly in tests.

### CODE REVIEW

File: test/lib/test_m3_server/test_m3_server.go [hunk 1]
```diff
@@ -141,18 +143,23 @@ func (m *FakeM3Service) GetMetrics() []*m3.Metric {
 // EmitMetricBatch is called by thrift message processor
 func (m *FakeM3Service) EmitMetricBatch(batch *m3.MetricBatch) (err error) {
 	m.lock.Lock()
+	defer func() { m.lock.Unlock() }()
```
Comment (ChuntaoLu) [inline test/lib/test_m3_server/test_m3_server.go]: why not just `defer m.lock.Unlock()`?

File: test/lib/test_m3_server/test_m3_server.go [hunk 2]
```diff
@@ -141,18 +143,23 @@ func (m *FakeM3Service) GetMetrics() []*m3.Metric {
...
+
 	m.batches = append(m.batches, batch)
 	if m.countBatches {
 		m.wg.Done()
 	}
 
 	for _, metric := range batch.Metrics {
-		m.metrics = append(m.metrics, metric)
+		tags := make(map[string]string)
```
Comment (ChuntaoLu) [inline test/lib/test_m3_server/test_m3_server.go]: nit: preallocate map with `len(tags)`

File: test/lib/test_m3_server/test_m3_server.go (hunk 1 above)
Comment (ravirajj) [reply test/lib/test_m3_server/test_m3_server.go]: Done.

File: test/lib/test_m3_server/test_m3_server.go (hunk 2 above)
Comment (ravirajj) [reply test/lib/test_m3_server/test_m3_server.go]: Done.

File: runtime/gateway.go [hunk 3]
```diff
@@ -57,23 +52,23 @@ type Options struct {
...
+	TChannelPort     int32
+	RealHTTPPort     int32
+	RealHTTPAddr     string
+	RealTChannelPort int32
+	RealTChannelAddr string
+	WaitGroup        *sync.WaitGroup
+	Channel          *tchannel.Channel
+	Logger           *zap.Logger
+	RootScope        tally.Scope
+	AllHostScope     tally.Scope
```
Comment (Raynos) [inline runtime/gateway.go]: breaking changes are expensive. we don't want to land breaking changes.

File: runtime/gateway.go [hunk 4]
```diff
@@ -322,24 +317,23 @@ func (gateway *Gateway) setupMetrics(config *StaticConfig) (err error) {
...
+		time.Duration(config.MustGetInt("metrics.flushInterval"))*time.Millisecond,
 	)
-	gateway.MetricsScope = gateway.RootScope.SubScope("all-workers")
-	gateway.PerWorkerMetricsScope = gateway.RootScope.SubScope("per-worker").Tagged(map[string]string{
-		"host": GetHostname(),
-	})
+	// As per M3 best practices, creating separate all-host and per-host metrics
+	// to reduce metric cardinality when querying metrics for all hosts.
+	gateway.AllHostScope = gateway.RootScope.Tagged(map[string]string{"host": "all"})
+	gateway.PerHostScope = gateway.RootScope.Tagged(map[string]string{"host": GetHostname()})
```
Comment (Raynos) [inline runtime/gateway.go]: We want the per host scope to be prefixed with some unique prefix to make grafana querying easier.

**Comment by Raynos** [comment]
See https://github.com/uber/zanzibar/pull/204

**Pull request merged**